teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,423 @@
+
+4. Secrets and RBAC concerns 
+
+    Today, if an interceptor needs access to a secret (e.g GitHub payload validation), the service account for the EventListener needs to get access to secrets. With the proposed model, this is no longer the case. Instead the GitHub interceptor deployment should use a role that has read access to secrets. This role would have to be a cluster role if it needs access to secrets across multiple namespaces. A security conscious admin might want to use [resourceNames](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources) to lock down the list of secrets the deployment has access to.
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+```
+<<[UNRESOLVED @dibyom]>>
+<<[/UNRESOLVED]>>
+```
+
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+
+```go
+	type InterceptorRequest struct {
+		// Body is the incoming HTTP event body
+		Body []byte `json:"body,omitempty"`
+		// Header are the headers for the incoming HTTP event
+		Header map[string][]string `json:"header,omitempty"`
+		// EventUrl is the URL of the incoming event
+		EventUrl string `json:"url,omitempty"`
+		// InterceptorParams are the user specified params for the interceptor
+		InterceptorParams map[string]string `json:"interceptor_ params,omitempty"`
+		// EventID is a unique ID assigned by Triggers to each event
+		EventID string `json:"event_id,omitempty"`
+		// TriggerID is of the form namespace/$ns/triggers/$name
+		TriggerID string `json:"trigger_id,omitempty"`
```

Comment:
Hey, that's a good idea. It also provides a nice buffer for adding additional context fields e.g. yesterday I realized that it might be useful to pass in the Trigger namespace too -- this way the interceptor can pull in the secrets from the trigger's namespace (like we do today.)

Reply:
ahh I see that your TriggerId already contains the trigger namespace :D